packages/search-in-workspace/src/browser/components/input-with-history.tsx

```diff
@@ -0,0 +1,131 @@
+import debounce = require('@theia/core/shared/lodash.debounce');
+
+interface HistoryState {
+    history: string[];
+};
+type InputAttributes = React.InputHTMLAttributes<HTMLInputElement>;
+
+export class InputWithHistory extends React.Component<InputAttributes, HistoryState> {
+    private input = React.createRef<HTMLInputElement>();
+
+    constructor(props: InputAttributes) {
+        super(props);
+        this.state = {
+            history: [],
+        };
+    }
+
+    componentDidUpdate(): void {
+        // Catch updates from the StatefulWidget restore method
+        if (this.value && !this.state.history.length) {
+            this.setHistory([this.value]);
+        }
+    }
+
+    private setHistory(history: string[]): void {
+        this.setState(prevState => ({
+            ...prevState,
+            history,
+        }));
+    }
+
+    get value(): string {
+        return this.input.current?.value ?? '';
+    }
+
+    set value(value: string) {
+        if (this.input.current) {
+            this.input.current.value = value;
+        }
+    }
+
+    /**
+     * Handle history navigation without overriding the parent's onKeyDown handler, if any.
+     */
+    protected readonly onKeyDown = (e: React.KeyboardEvent<HTMLInputElement>): void => {
+        if (Key.ARROW_UP.keyCode === KeyCode.createKeyCode(e.nativeEvent).key?.keyCode) {
+            e.preventDefault();
+            this.previousValue();
+        } else if (Key.ARROW_DOWN.keyCode === KeyCode.createKeyCode(e.nativeEvent).key?.keyCode) {
+            e.preventDefault();
+            this.nextValue();
+        }
+        this.props.onKeyDown?.(e);
+    };
+
+    /**
+     * Switch the input's text to the previous value, if any.
+     */
+    previousValue(): void {
+        const { history } = this.state;
+        const index = history.indexOf(this.value);
+        if (index > 0 && index < history.length) {
+            this.value = history[index - 1];
+        }
+    }
+
+    /**
+     * Switch the input's text to the next value, if any.
+     */
+    nextValue(): void {
+        const { history } = this.state;
+        const index = history.indexOf(this.value);
+        if (index >= 0 && index < history.length - 1) {
+            this.value = history[index + 1];
+        }
+    }
+
+    /**
+     * Handle history collection without overriding the parent's onChange handler, if any.
+     */
+    protected readonly onChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
+        this.addToHistory();
+        this.props.onChange?.(e);
+    };
+
+    /**
+     * Add a nonempty current value to the history, if not already present. (Debounced, 1 second delay.)
+     */
+    readonly addToHistory = debounce(this.doAddToHistory, 1000);
+
+    private doAddToHistory(): void {
+        if (!this.value) {
+            return;
+        }
+        const { history } = this.state;
+        const index = history.indexOf(this.value);
```

Comment:
VSCode allows repetitions in the history, so this strategy, which doesn't allow duplication, may need to be adjusted to align with that behavior. It may also be desirable to have a maximum history length, though I haven't checked whether VSCode has one or what it might be.

Reply:
VSCode [uses a Set](https://github.com/microsoft/vscode/blob/d2854fd7877a1214a5e6d16315b77df7b3cf206e/src/vs/base/common/history.ts#L24) instead of an array to prevent repetition/duplication of history elements, and a maximum history length of [100](https://github.com/microsoft/vscode/blob/d2854fd7877a1214a5e6d16315b77df7b3cf206e/src/vs/base/browser/ui/inputbox/inputBox.ts#L636) entries

Reply:
They do use a set, but they also have a mechanism for [tracking position](https://github.com/microsoft/vscode/blob/d2854fd7877a1214a5e6d16315b77df7b3cf206e/src/vs/base/common/history.ts#L30-L35) in the history so that the same element can appear at different places is the history.

Reply:
Nevermind. VSCode's behavior was more subtle than I had appreciated. Always dangerous making pronouncements when timeouts are invoved...